File: Data_Structures_And_Algorithms-UC_SanDiego/1-Algorithmic_Toolbox/Week5/Programming_Assignments/1_money_change_again/change_dp.cpp
#include <iostream>
#include <limits>
#include <vector>

int get_change(int m) {
  int element = std::numeric_limits<int>::max();
  std::vector<int> coins {1, 3, 4};
  std::vector<int> minCoins {};
  int sumCoins = 0;
  for (int i=1; i!=m+2; i++){
    minCoins.push_back(element);
  }
  minCoins[0] = 0;
  for (int i=1; i!=m+1; i++){
    for (auto el:coins){
      if (i>=el){
        sumCoins = minCoins[i-el] + 1;
        if (sumCoins<minCoins[i]){
          minCoins[i] = sumCoins;
        }
      }
    }
  }
  return minCoins[m];
}

int main() {
  int m;
  std::cin >> m;
  std::cout << get_change(m) << '\n';
  return 0;
}

File: Data_Structures_And_Algorithms-UC_SanDiego/1-Algorithmic_Toolbox/Week5/Programming_Assignments/1_money_change_again/change_dp.py
# Uses python3
import sys

def get_change(m):
    element = float("inf")
    minCoins = list()
    sumCoins = 0
    coins = [1,3,4]
    minCoins[0] = 0
    for i in range(1, m+2):
        minCoins.append(element)
    for i in range(1, m + 1):
        for el in coins:
            if i >= el:
                sumCoins = minCoins[i - el] + 1
                if sumCoins < minCoins[i]:
                    minCoins[i] = sumCoins
    return minCoins[money]

if __name__ == '__main__':
    m = int(sys.stdin.read())
    print(get_change(m))
